pkg/common/cliprinter/cliprinter.go

```diff
@@ -0,0 +1,131 @@
+package cliprinter
+
+import (
+	"io"
+	"os"
+	"sync"
+
+	"github.com/spiffe/spire/pkg/common/cliprinter/errorjson"
+	"github.com/spiffe/spire/pkg/common/cliprinter/errorpretty"
+	"github.com/spiffe/spire/pkg/common/cliprinter/protojson"
+	"github.com/spiffe/spire/pkg/common/cliprinter/protopretty"
+	"github.com/spiffe/spire/pkg/common/cliprinter/structjson"
+	"github.com/spiffe/spire/pkg/common/cliprinter/structpretty"
+	"google.golang.org/protobuf/proto"
+)
+
+type Printer interface {
+	PrintError(error)
+	PrintProto(...proto.Message)
+	PrintStruct(...interface{})
+}
+
+type DefaultPrinter struct {
+	format formatType
+	mtx    sync.RWMutex
+
+	stdout io.Writer
+	stderr io.Writer
+
+	customPrettyFunc func(...interface{})
+}
+
+// NewDefaultPrinter returns a new printer configured with
+// the default formatter
+func NewDefaultPrinter() *DefaultPrinter {
+	return newPrinter(pretty)
+}
+
+// NewPrinter returns a new printer and configures it to use
+// the specified formatter
+func NewPrinter(format string) (*DefaultPrinter, error) {
+	f, err := strToFormatType(format)
+	if err != nil {
+		return nil, err
+	}
+
+	return newPrinter(f), nil
+}
+
+func newPrinter(f formatType) *DefaultPrinter {
+	return &DefaultPrinter{
+		format: f,
+		stdout: os.Stdout,
+		stderr: os.Stderr,
+	}
+}
+
+// PrintError prints an error and applies the configured formatting
+func (p *DefaultPrinter) PrintError(err error) {
+	switch p.format {
+	case json:
+		errorjson.Print(err, p.stdout, p.stderr)
+	case pretty:
+	default:
+		p.printPrettyError(err, p.stdout, p.stderr)
+	}
+}
+
+// PrintError prints an error and applies the configured formatting
+func (p *DefaultPrinter) PrintProto(msg ...proto.Message) {
+	switch p.format {
+	case json:
+		protojson.Print(msg, p.stdout, p.stderr)
+	case pretty:
+	default:
+		p.printPrettyProto(msg, p.stdout, p.stderr)
+	}
+}
+
+// PrintStruct prints a struct and applies the configured formatting
+func (p *DefaultPrinter) PrintStruct(msg ...interface{}) {
+	switch p.format {
+	case json:
+		structjson.Print(msg, p.stdout, p.stderr)
+	case pretty:
+	default:
+		p.printPrettyStruct(msg, p.stdout, p.stderr)
+	}
+}
+
+// SetCustomPrettyPrinter can be used to override the pretty print logic
+// that normally ships with this package. Its intended use is to allow
+// for the use of cliprinter while still retaining backwards compatibility
+// with legacy/bespoke pretty print output.
+func (p *DefaultPrinter) SetCustomPrettyPrinter(f func(...interface{})) {
+	p.mtx.Lock()
```

Comment:
Are we anticipating the custom printer changing at runtime across multiple goroutines? If not, I wonder if this mutex raises more questions than problems it fixes...